Proposal contribution matching

AMStrix opened the pull request:
Closes #88.

I welcome feedback on the UX, I like the ribbon. But I'm not sure I love the `CampaignBlock` "x2", might be too subtle. But maybe it's ok? 

I suspect we might want to have more control/definition on this feature's behavior, but I think it wise to have a final pass at that once refunds, payout & staking basics are also present.

### What this does
- adds `contribution_matching` float to `Proposal` model
- derives funded amount on the frontend using `contributionMatching`
- limits its value to 1 (on) or 0 (off), allows arbitrary multipliers in the future
- admin: adds ability to set matching status 
- frontend: ribbon visual cue for proposal list cards with matching turned on
- frontend: subtle callout in `CampaignBlock` for matching

### What this doesn't do
- limit when a proposal can have its matching state set
- limit the amount a user can donate based on multiplier (user can contribute the full remaining amount if they want)
- display proposal accounting for admin to easily find out how much is owed via matching

### Testing
- upgrade your database, this should add the column and set the `contribution_matching` default to '0' (via `server_default`)
- use admin to set any proposal to "matching" (switch under delete button)
- make sure the proposal is live
- make sure the card has the ribbon callout
- make sure the proposal campaign block has the "x2" on the "Funding" line, should have a tooltip
- contributions to the proposal should be doubled (on list, detail and profile > created)
- turn on matching for a contribution with > 50% actual funding, make sure it is listed as fully funded, and the amount funded is equal to the target and not greater than the target.

## Screens
![screen shot 2019-01-28 at 2 33 38 pm](https://user-images.githubusercontent.com/11430954/51866346-6440e680-230e-11e9-8a5e-88d75bd2fd1c.png)

![screen shot 2019-01-28 at 2 41 09 pm](https://user-images.githubusercontent.com/11430954/51866358-699e3100-230e-11e9-9e01-c7ee9aa684e7.png)

![screen shot 2019-01-28 at 2 37 17 pm](https://user-images.githubusercontent.com/11430954/51866366-6e62e500-230e-11e9-994a-5f4480c86e5a.png)

dternyak commented on admin/src/components/ProposalDetail/index.tsx:
```diff
@@ -42,12 +54,48 @@ class ProposalDetailNaked extends React.Component<Props, State> {
+            <div>Turn {p.contributionMatching ? 'off' : 'on'} contribution matching?</div>
+            {p.status === PROPOSAL_STATUS.LIVE && (
+              <div>
+                This is a LIVE proposal, this will alter the funding state of the
```
Interesting! Trying to think this through. What happens if a proposal is > 50% funded before matching is turned on?

AMStrix replied:
It will appear fully funded on frontend, of course the amount being matched will be < 50%, which seems like it would be ok provided we give a nice accounting to admins.

AMStrix replied:
The worst case I can think of is turning off matching on a proposal @ > 50% funding. Users would have received funded notifications and be under the impression that they can start work and request payouts.

dternyak commented:
Definitely giving this a full run through later.

On the design side, I do think the x2 in the campaign block may be a bit subtle. What if we made it a banner on the campaign block as well as in the listview cards? That should give enough room to describe it as "Matched 2x" or something of the like.

dternyak commented:
> limit the amount a user can donate based on multiplier (user can contribute the full remaining amount if they want)

This makes sense, considering users can fund a proposal however much they want at the protocol level.

We should consider how we'll display over-funding in general, since this is something that *will* happen in practice.

AMStrix commented on admin/src/components/ProposalDetail/index.tsx:
```diff
@@ -42,12 +54,48 @@ class ProposalDetailNaked extends React.Component<Props, State> {
+        cancelText="cancel"
+      >
+        <div className="ProposalDetail-controls-control">
+          <Switch checked={p.contributionMatching === 1} loading={false} />{' '}
```
Oh, good point!

AMStrix commented on frontend/client/utils/api.ts:
```diff
@@ -11,9 +11,20 @@ export function formatUserForPost(user: User) {
+  const targetZat = toZat(p.target);
+  // funded = funded + funded * matching
+  let fundedZat = toZat(p.funded).muln(1 + p.contributionMatching);
+  // if funded exceeds the target, due to matching, just return the target
```
Yeah, one SoT. I'd like full info on the model tho.

Any thoughts on field names? `contributed` (sum of contributions) & `funded` (derived) maybe?

wbobeirne replied:
SGTM 👍

AMStrix commented:
Ok, this should be good to go.
* added `contributed` (sum of contributions) field to `Proposal` as hybrid property
* changed `funded` to include matching funds
* made `funded` a hybrid property
* improved the campaign block UX per @wbobeirne 's excellent advice/code
* fixed popover confirm target on admin proposal detail matching switch
* adjusted branching migration (post merge)

wbobeirne commented:
You're a prince. Hybrid properties are cool. Let's merge this bad boy in.
